sky/data/data_utils.py

```diff
@@ -69,3 +78,118 @@ def is_cloud_store_url(url):
+    """
+    grouped_files = {}
+    dirs = []
+    for source in source_list:
+        source = os.path.abspath(os.path.expanduser(source))
+        if os.path.isdir(source):
+            dirs.append(source)
+        else:
+            base_path = os.path.dirname(source)
+            file_name = os.path.basename(source)
+            if base_path not in grouped_files:
+                grouped_files[base_path] = []
+            grouped_files[base_path].append(file_name)
+    return grouped_files, dirs
+
+
+def parallel_upload(source_path_list: List[Path],
+                    filesync_command_generator: Callable,
+                    dirsync_command_generator: Callable,
+                    bucket_name: str,
+                    access_denied_message: str,
+                    create_dirs: bool = False,
+                    max_concurrent_uploads: Optional[int] = None) -> None:
+    """Helper function to run parallel uploads for a list of paths.
+
+    Used by S3Store and GCSStore to run rsync commands in parallel by
+    providing appropriate command generators.
+
+    Args:
+        source_path_list: List of paths to local files or directories
+        filesync_command_generator: Callable that generates rsync command
+            for a list of files belonging to the same dir.
+        dirsync_command_generator: Callable that generates rsync command
+            for a directory.
+        access_denied_message: Message to intercept from the underlying
+            upload utility when permissions are insufficient. Used in
+            exception handling.
+        create_dirs: If the local_path is a directory and this is set to
+            False, the contents of the directory are directly uploaded to
+            root of the bucket. If the local_path is a directory and this is
+            set to True, the directory is created in the bucket root and
+            contents are uploaded to it.
+        max_concurrent_uploads: Maximum number of concurrent threads to use
+            to upload files.
+    """
+    # Generate gsutil rsync command for files and dirs
+    commands = []
+    grouped_files, dirs = group_files_by_dir(source_path_list)
+    # Generate file upload commands
+    for dir_path, file_names in grouped_files.items():
+        sync_command = filesync_command_generator(dir_path, file_names)
+        commands.append(sync_command)
+    # Generate dir upload commands
+    for dir_path in dirs:
+        if create_dirs:
+            dest_dir_name = os.path.basename(dir_path)
+        else:
+            dest_dir_name = ''
+        sync_command = dirsync_command_generator(dir_path, dest_dir_name)
+        commands.append(sync_command)
+
+    # Run commands in parallel
+    with pool.ThreadPool(processes=max_concurrent_uploads) as p:
+        p.starmap(
+            run_upload_cli,
+            zip(commands, [access_denied_message] * len(commands),
+                [bucket_name] * len(commands)))
```

Comment:
Will this correctly fail the other threads when one child fails? Just want to make sure that there will be no leaked thread.
Here is the reason we switched to `p.imap` in `subprocess_utils`.
https://stackoverflow.com/questions/25790279/python-multiprocessing-early-termination